### PR TITLE
fix(mongo_client): remove validateOptions

### PR DESCRIPTION
**Pull request by kvwalker**
Fixes NODE-1754

### CODE REVIEW

**Comment by kvwalker** [comment]
I think this would be a breaking change and I don't know if people actually use this feature. An alternative to this solution is to set `optionsValidationLevel` to `error` if `validateOptions` is `true`. That way, we can still remove the `validOptions` function and keep the existing behavior.

**Comment by mbroadst** [comment]
@kvwalker @daprahamian could we potentially just map this to the `optionsValidationLevel`? Agree that it's potentially breaking, but then you might argue that options management as a whole is potentially breaking. Maybe we should defer to meeting on Thursday to discuss this?

File: lib/options_validator.js [hunk 1]
```diff
@@ -64,10 +64,10 @@ function assertArity(args, requiredArity) {
  * @param {object} [validationOptions] Options for the validation itself.
  * @param {Logger} [validationOptions.logger] A logger instance to use if validation fails.
  * @param {string} [validationOptions.optionsValidationLevel] The level at which to validate the providedOptions ('none', 'warn', or 'error').
- * @param {boolean} [warnOnUnknownOptions=false] If true, warn on any option that is not in the schema.
+ * @param {string} [validationOptions.unknownOptionsWarningLevel] The level at which to handle unknown options ('none', 'warn', or 'error').
```
Comment (kvwalker) [inline lib/options_validator.js]: The changes in `options_validator` supersede the changes in #1898.

File: lib/mongo_client.js [hunk 2]
```diff
@@ -219,9 +219,15 @@ function MongoClient(url, options) {
...
+
+  let unknownOptionsWarningLevel = 'warn';
+  if (options.validateOptions) {
+    unknownOptionsWarningLevel = 'error';
+  }
+
   validate(connectSchema, options, {
     optionsValidationLevel: this.optionsValidationLevel,
-    warnOnUnknownOptions: true
+    unknownOptionsWarningLevel: unknownOptionsWarningLevel
```
Comment (mbroadst) [inline lib/mongo_client.js]: you can just do `{ unknownOptionsWarningLevel }`

File: lib/mongo_client.js [hunk 3]
```diff
@@ -531,7 +534,6 @@ MongoClient.prototype.isConnected = function(options) {
  * @param {boolean} [options.promoteLongs=true] Promotes long values to number if they fit inside the 53 bits resolution
  * @param {boolean} [options.domainsEnabled=false] Enable the wrapping of the callback in the current domain, disabled by default to avoid perf hit
  * @param {boolean|function} [options.checkServerIdentity=true] Ensure we check server identify during SSL, set to false to disable checking. Only works for Node 0.12.x or higher. You can pass in a boolean or your own checkServerIdentity override function
- * @param {object} [options.validateOptions=false] Validate MongoClient passed in options for correctness
```
Comment (mbroadst) [inline lib/mongo_client.js]: I think we still need to document this, even if we are deprecating it. You support it's value below.

File: lib/mongo_client.js [hunk 4]
```diff
@@ -84,8 +83,8 @@ const connectSchema = {
   promoteBuffers: { type: 'boolean' },
   promoteLongs: { type: 'boolean' },
   domainsEnabled: { type: 'boolean', default: false },
+  validateOptions: { type: 'boolean', default: false, deprecated: 'unknownOptionsWarningLevel' },
```
Comment (mbroadst) [inline lib/mongo_client.js]: I expected this to read:
```
validateOptions: { type: 'boolean', default: false, deprecated: true },
```
Why is there a string `unknownOptionsWarningLevel` here?

File: test/functional/mongo_client_options_tests.js [hunk 5]
```diff
@@ -74,7 +74,7 @@ describe('MongoClient Options', function() {
   /**
    * @ignore
    */
-  it('should error on unexpected options', {
+  it.skip('should error on unexpected options', {
```
Comment (mbroadst) [inline test/functional/mongo_client_options_tests.js]: 🤔 does this mean our mapping doesn't really work? Why is this test skipped?

File: test/functional/mongo_client_tests.js [hunk 6]
```diff
@@ -792,4 +797,27 @@ describe('MongoClient', function() {
...
   });
+
+  it('Should set optionsValidationLevel when validateOptions is set', {
+    metadata: { requires: { topology: ['single'] } },
+    // The actual test we wish to run
+    test: function() {
+      const configuration = this.configuration;
+      try {
+        configuration.newClient({}, { validateOptions: true, unknownOption: true });
+      } catch (err) {
```
Comment (mbroadst) [inline test/functional/mongo_client_tests.js]: to me it looks like if `validateOptions` is true, the level should be [warn](https://github.com/mongodb/node-mongodb-native/pull/1895/files#diff-12bc6199a1348e0428b4b9ebeb881206L654) not `error`

File: lib/mongo_client.js (hunk 4 above)
Comment (kvwalker) [reply lib/mongo_client.js]: The string was @daprahamian's suggestion, I believe, so we can emit a deprecation notice telling users that `validateOptions` is deprecated and they should use `unknownOptionsWarningLevel` instead.

File: test/functional/mongo_client_tests.js (hunk 6 above)
Comment (kvwalker) [reply test/functional/mongo_client_tests.js]: If `validateOptions` is true, [the level should be error.](https://github.com/mongodb/node-mongodb-native/pull/1895/files#diff-12bc6199a1348e0428b4b9ebeb881206L651) The warning is emitted by default.

File: test/functional/mongo_client_options_tests.js (hunk 5 above)
Comment (kvwalker) [reply test/functional/mongo_client_options_tests.js]: Okay, I've fixed things now so the unsupported options message is consistent and options will only be validated on calls to `connect`. `validateOptions` can still be set at the constructor level (as it previously could be), but we will only warn about unknown options during `connect`. This maintains consistency with the existing behavior of the driver.

File: lib/mongo_client.js [hunk 7]
```diff
@@ -285,14 +305,22 @@ Object.defineProperty(MongoClient.prototype, 'optionsValidationLevel', {
  * @return {Promise<MongoClient>} returns Promise if no callback passed
  */
 MongoClient.prototype.connect = function(callback) {
-  // Validate options object
-  const err = validOptions(this.s.options);
+  try {
```
Comment (kvwalker) [inline lib/mongo_client.js]: This is necessary to pass the validation error to the callback and ensures that the `should error on unsupported options` test passes.

**Pull request merged**